### PR TITLE
Async utils

### PR DESCRIPTION
**Pull request by Hugheth**

### CODE REVIEW

File: lib/ClassUtils.lua [hunk 1]
```diff
@@ -24,11 +24,15 @@ function ClassUtils.makeClass(name, constructor)
     return Class
 end
 
-function ClassUtils.makeWrapperClass(name)
+function ClassUtils.makeConstructedClass(name, constructor)
```
Comment (ckdavid) [inline lib/ClassUtils.lua]: Aren't these changes unrelated to the async stuff?

File: lib/AsyncUtils.lua [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+				delayInSeconds,
+				function()
+					resolve()
+				end
+			)
+		end
+	)
+end
+
+function AsyncUtils.wrapAsync(fn)
```
Comment (inversion) [inline lib/AsyncUtils.lua]: can we get a doc comment

File: lib/AsyncUtils.lua [hunk 3]
```diff
@@ -0,0 +1,131 @@
+local TableUtils = require(script.Parent.TableUtils)
+local Promise = require(script.Parent.Parent.Promise)
+local AsyncUtils = {}
+
+local baseRandomStream = Random.new()
+
+function AsyncUtils.parallel(things)
```
Comment (inversion) [inline lib/AsyncUtils.lua]: how does this work wrt error handling?

File: lib/AsyncUtils.lua [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+			backoffOptions.onFail(response)
+			return Promise.reject(response)
+		else
+			return backoffThenRetry(response)
+		end
+	elseif not Promise.is(response) then
+		backoffOptions.onDone(response, getDurationMs())
+		return Promise.resolve(response)
+	elseif backoffOptions.maxTries == 1 then
+		return response:catch(
```
Comment (inversion) [inline lib/AsyncUtils.lua]: onDone not called?

File: lib/AsyncUtils.lua [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+					else
+						reject(result)
+					end
+				end
+			)()
+		end
+	)
+end
+
+function AsyncUtils.retryWithBackoff(getPromise, backoffOptions)
```
Comment (inversion) [inline lib/AsyncUtils.lua]: this stuff could be tested within the new studio testez harness

Comment (inversion) [inline lib/AsyncUtils.lua]: Documentation for options?

File: lib/AsyncUtils.lua [hunk 6]
```diff
@@ -0,0 +1,131 @@
+local TableUtils = require(script.Parent.TableUtils)
```
Comment (inversion) [inline lib/AsyncUtils.lua]: None of these methods have much in the way of assertions to validate parameters

File: lib/ClassUtils.lua [hunk 7]
```diff
@@ -25,16 +25,20 @@ function ClassUtils.makeClass(name, constructor)
 end
 
 function ClassUtils.makeConstructedClass(name, constructor)
```
Comment (inversion) [inline lib/ClassUtils.lua]: makeConstructedClass and makeClassWithConstructor

File: lib/AsyncUtils.lua (hunk 3 above)
Comment (Hugheth) [reply lib/AsyncUtils.lua]: falls back to usual Promise.all behaviour which is fail fast and once if any child fails

File: lib/ClassUtils.lua (hunk 7 above)
Comment (Hugheth) [reply lib/ClassUtils.lua]: The point is a "constructed" class is one where the instance is already created with the fields passed in to new before the constructor is called. I'll update the "constructor" parameter name and documentation to explain

File: lib/ClassUtils.lua (hunk 1 above)
Comment (Hugheth) [reply lib/ClassUtils.lua]: Yes, please can we merge with this change to fix tests on roblox as that's dependent on this branch, and I'll apply the proposed change immediately after

**Pull request merged**